### ISSUE TITLE
Recurrent bug with incorrectly calculated user upload limits / Повторяющаяся ошибка с некорректным расчетом лимита загрузок

### ISSUE THREAD

**Issue by paul-k-pastvu** [PastVu/pastvu#342]
См. сообщение https://pastvu.com/news/153?hl=comment-2260284
Участник: https://pastvu.com/u/Pirogov/manage

![image](https://user-images.githubusercontent.com/60163100/102676139-2de40d00-41ad-11eb-9e18-59a882fcc424.png)

![image](https://user-images.githubusercontent.com/60163100/102676181-58ce6100-41ad-11eb-82ea-6b8b6d9ad378.png)

**Comment by paul-k-pastvu** [comment]
@klimashkin что-то вручную делал в таких случаях - то ли обновлял статистику, то ли перезапускал апп.
Но нужно уже однажды разобраться, что эе происходит, и починить.

**Comment by anvaruk** [comment]
нашел старую инструкцию для пересчета

```
mongo
use pastvu
db.loadServerScripts()
calcUserStats()
```

**Comment by paul-k-pastvu** [comment]
Может, нужно это дело раз в сутки запускать, по расписанию?

**Comment by aeifn** [comment]
> Может, нужно это дело раз в сутки запускать, по расписанию?

я запустил один раз, стало лучше?

**Comment by kabalin** [comment]
По идее само должно работать:
https://github.com/PastVu/pastvu/blob/master/app.js#L317

> Может, нужно это дело раз в сутки запускать, по расписанию?

[Раз в двое](https://github.com/PastVu/pastvu/blob/master/controllers/_session.js#L784) запускается на данный момент. @aeifn ошибок в логах нет?

**Comment by aeifn** [comment]
> По идее само должно работать:
> https://github.com/PastVu/pastvu/blob/master/app.js#L317
> 
> > Может, нужно это дело раз в сутки запускать, по расписанию?
> 
> [Раз в двое](https://github.com/PastVu/pastvu/blob/master/controllers/_session.js#L784) запускается на данный момент. @aeifn ошибок в логах нет?

ошибок в логах primary-приложения нету.

**Comment by kabalin** [comment]
> ошибок в логах primary-приложения нету.
@paul-k-pastvu сейчас как выглядит, исправилось после ручного запуска? Честно сказать, я не до конца уверен, что проблема в задаче пересчета. По идее, текущая статистика в большинстве случаев пересчитывается по мере изменений влияющих на нее (https://github.com/PastVu/pastvu/blob/master/controllers/photo.js#L942).

**Comment by paul-k-pastvu** [comment]
Да, в тот раз исправилось. Вопрос, почему время от время от времени сбоит (это далеко не первый раз). 
Вопрос несрочный, раз лечится пересчетом, но интересный.

**Comment by kabalin** [comment]
Я думаю, что дело в масштабировании приложения. Я как раз с этой функцией разбирался в процессе изменений для Monogo 4.  `calcUserStatsJob` запускается только на "праймари", у этой задачи есть две части - одна обновляет статистику в базе (через dbEval вызов), [вторая часть](https://github.com/PastVu/pastvu/blob/master/controllers/_session.js#L789) обновляет статистику у пользователей с открытыми сессиями. Вот эта вторая часть тоже запускается только на праймари, т.е. срабатывает не для всех пользователей. С функцией, которая пересчитывает в режиме реального времени, аналогичная ситуация - в базу она изменения внесет, но [та часть](https://github.com/PastVu/pastvu/blob/master/controllers/photo.js#L944) которая должна обновить текущего пользователя сработает только в случае если модератор и пользователь находятся в момент изменения статуса снимка на одном фронтэнде.

В качестве временного решения - если пользователь разлогинится и зайдет снова - статистика будет верной.

**Comment by aeifn** [comment]
> Я думаю, что дело в масштабировании приложения. Я как раз с этой функцией разбирался в процессе изменений для Monogo 4. `calcUserStatsJob` запускается только на "праймари", у этой задачи есть две части - одна обновляет статистику в базе (через dbEval вызов), [вторая часть](https://github.com/PastVu/pastvu/blob/master/controllers/_session.js#L789) обновляет статистику у пользователей с открытыми сессиями.

А как этот вопрос был решен на английской версии?


> Вот эта вторая часть тоже запускается только на праймари, т.е. срабатывает не для всех пользователей.

Вернее, ни у кого, т. к. инстанс primary никак не выпущен наружу.

**Comment by aeifn** [comment]
> Я думаю, что дело в масштабировании приложения.

Да, но это самое реплицирование сейчас обеспечивает по крайней мере
- стабильную работу без зависаний и без необходимости вручную что-то администрировать
- возможность обновлений без остановки приложения

Пользовательские инстансы должны быть подписаны на некоторую IPC-шину, и обновлять состояние при получении сервисных сообщений. Следуя принципу Оккама, [можно вообще использовать redis в качестве брокера](https://redis.io/topics/pubsub).

**Comment by kabalin** [comment]
> А как этот вопрос был решен на английской версии?

Как я понимаю там `calcUserStatsJob` до https://github.com/PastVu/pastvu/commit/61756231a0f6d4aa353689816c2842d76efab50b работал так же как и на мастере, т.е. задача в полном объеме запускалась для каждого инстанса.

> Да, но это самое реплицирование сейчас обеспечивает по крайней мере

Это бесспорно.

> Пользовательские инстансы должны быть подписаны на некоторую IPC-шину, и обновлять состояние при получении сервисных сообщений.

Все верно, так примерно и будет ([sneak peek](https://github.com/kabalin/pastvu/blob/mongo4-p3/app.js#L315)) 😉

**Comment by kabalin** [comment]
> В качестве временного решения - если пользователь разлогинится и зайдет снова - статистика будет верной.

Похоже я ошибался, в качестве временного решения это не подходит. Хотя я смог тогда воспроизвести ситуацию где обновление счетчика не происходило как я описал в комментарии выше, по всей видимости проблема в чем-то другом.

**Comment by anvaruk** [comment]
вручную пока лимит увеличить? 
![image](https://user-images.githubusercontent.com/2795267/137287924-8157badf-44d7-41c4-85c8-4dcb8d40b856.png)

**Comment by aeifn** [comment]
> На системный крон выносить мне кажется смысла пока нет (но если так удобнее, то можно, только потом нужно будет не забыть убрать после апгрейда).

Ну если что, системный крон у нас управляется через https://github.com/pastvu/ansible

**Comment by kabalin** [comment]
Скорее всего проблема будет окончательно решена в #435
